### DISCUSSION TITLE
☂️ Various features that exist in DDA but do not exist in BN

### DISCUSSION THREAD

**Discussion by Firestorm01X2** [cataclysmbn/Cataclysm-BN#3872] [General]
**Tekker#7841** prepared list of thing that exist in DDA, but does not exist in BN and vice versa.
I am creating issue because some of this thing are relatively small and it is easy to miss it.
It is good idea to review that list and port some things.

**Exist in DDA but not in BN:**

- Layered inventory **(For now we prefer to avoid that)**
- Food spoilage factors in thaw times, making food last ages longer **(BN has that)**
- Enemy display above minimap shows numbers of each enemy type **(Definitely interesting QoL feature)**
- More car part tiles **( Possible list of PR is here: https://github.com/cataclysmbnteam/Cataclysm-BN/pull/418 )**
- Can take follower equiped items from trade menu **(Nice QoL)**
- Cracker sandwiches (**Sounds minor, is it really provide something unique? But still.**)
- Items and enemies show what mod they're from in descriptions **Ported**
- Books are color coded by skill/fun **Ported**
- Saved character presets remember what their name was **(Nice QoL)**

**Exist in BN but non in DDA:**

- Watches/sheathes/holsters automatically equip to empty side
- Read books from the ground
- Power grid
- 'E' on counters doesn't prompt crafting
- Can still use 'E' to loot **(Already restored in DDA?)**
```[tasklist]
### Tasks
- [ ] https://github.com/cataclysmbnteam/Cataclysm-BN/issues/2389
```

**Comment by TekkerSkillet** [comment]
in BN food is frozen at 32.
in DDA food is frozen at 32.
in BN food beings to rot at 33.
in DDA food is still frozen at 33 and stays that way for hours of "defrosting" time, letting it stay frozen indefinitely if the temps change back to 32 before it has thawed.

That means food in DDA, on a spring 0 start, stays frozen and not rotting for 20+ days longer than in BN.

So no, that's not already in BN.

Additional things for the list because I don't know how to use github:
DDA has 
-Ladders/Stairs that go up/down on the same tile, meaning you don't have to walk to a separate ladder/stairs to climb things like radio towers and silos, you just hit up/down again.
-Keeping the Syringe from used Flu shots
-Cat Ear accessories don't take up a head slot, blocking other head wear
-Different bleed levels indicated by different coloring of the LMB, pink being light bleeding, darker reds being progressively worse bleeding.

BN has
-Showing your carrying capacity in the general HUD

**Comment by Zireael07** [comment]
BN has: foraging yielding harvest materials https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1134

**Comment by chaosvolt** [comment]
My current list of things to want to mess with at some point:
* Weapon categories for mod intercompatibility (https://github.com/CleverRaven/Cataclysm-DDA/pull/50212, https://github.com/CleverRaven/Cataclysm-DDA/pull/52402, follow-up PRs that added wcategories to styles and weapons).
* Ability to refuel reloadable terrain/furniture (smokers and such) using adjacent fuel (https://github.com/CleverRaven/Cataclysm-DDA/pull/51442).
* Leash and lead tamed animals (https://github.com/CleverRaven/Cataclysm-DDA/pull/50890).
* Allowing effects to block regen (https://github.com/CleverRaven/Cataclysm-DDA/pull/50439), needs an actual implementation instead of source PR's half-assing it to only give it to debug monster.
* Steal the mana modifier thing for Olanti (part of https://github.com/CleverRaven/Cataclysm-DDA/pull/44357).
* Allowing NPCs to toggle on items in combat (https://github.com/CleverRaven/Cataclysm-DDA/pull/52259), it's a new-ish feature so will need double-checking for bugs.
* Allowing healing rate changes for any effect, not just hardcoded to bandaged and disinfected effects (https://github.com/CleverRaven/Cataclysm-DDA/pull/40072).
* Maybe JSONization of parking lots (https://github.com/CleverRaven/Cataclysm-DDA/pull/44933). Problem is, like we found out with mine entrances (and earlier, industrial centers), if generic parking lots are being used in an overmap special in a way where the hardcoded placement would break, chances are a single generic parking center layout isn't useful to us in that context, and it's probably better to make a custom om_terrain that fits that area instead.

Ballistic damage has since been ported over and is now removed from the wishlist.

**Comment by chaosvolt** [comment]
Another for the list: JSONized pet-taming items. https://github.com/CleverRaven/Cataclysm-DDA/pull/50948

**Comment by chaosvolt** [comment]
Also potentially useful: https://github.com/CleverRaven/Cataclysm-DDA/pull/53605

Question for @Ramza13 (the PR OP) though, how easy would it be to also implement the same variety to ter-turn-transform, which currently only allows terrain and furniture? Being able to look for a specific trap (or even field) and convert it to another would be extremely handy.

An alternative could be `translate_ter` feature, which is presently limited solely to terrain.

**Comment by chaosvolt** [comment]
We also keep running into cases where transforming mutations would be handy:
https://github.com/CleverRaven/Cataclysm-DDA/pull/39523
https://github.com/CleverRaven/Cataclysm-DDA/pull/40803
https://github.com/CleverRaven/Cataclysm-DDA/pull/41448

**Comment by Ramza13** [comment]
> Also potentially useful: [CleverRaven/Cataclysm-DDA#53605](https://github.com/CleverRaven/Cataclysm-DDA/pull/53605)
> 
> Question for @Ramza13 (the PR OP) though, how easy would it be to also implement the same variety to ter-turn-transform, which currently only allows terrain and furniture? Being able to look for a specific trap (or even field) and convert it to another would be extremely handy.
> 
> An alternative could be `translate_ter` feature, which is presently limited solely to terrain.

Hmm adding the ability to transform transform fields and traps sounds interesting.  Might help me make my eventual nether dungeon more dynamic.  I haven't looked much at the code but I don't think that sounds too hard.  I'll add it to my to-do list. Edit: Just to be clear I am intending to add these things to DDA not BN but they should hopefully be easy to merge.

**Comment by chaosvolt** [comment]
Will be neat to see if you do tackle this, and feel free to link to a PR if you end up tackling it, might see if I can port it myself sometime.

EDIT: transforming traps in particular would be a good expansion of that fatigue-eating feature in the above PR in fact, as then it'd fully mimic the effects of ARTC_PORTAL, able to affect portal traps in addition to fatigue fields.

**Comment by Ramza13** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/pull/54491
This should handle trap and field transforms.

**Comment by chaosvolt** [comment]
Oooh, nice. I might set aside time this week to work on some of the stuff on my list of things I want to mess with porting over.

**Comment by UmbralReaper** [comment]
As of https://github.com/CleverRaven/Cataclysm-DDA/pull/54216, DDA has restored negative morale for killing child zombies.

**Comment by chaosvolt** [comment]
Last I checked this feature is working in BN, so the fix is for a DDA-specific bug.

**Comment by UmbralReaper** [comment]
> Last I checked this feature is working in BN, so the fix is for a DDA-specific bug.

I was referring to this dotpoint in the opening post

> Killing child zombies doesn't give negative morale (Probably it is better to keep morale effect- it makes child zombies somewhat unique)

**Comment by chaosvolt** [comment]
This has no relevance to BN currently, you're talking about a feature that was only broken in DDA:
![image](https://user-images.githubusercontent.com/11582235/149721729-bf150672-a52f-42fe-a9c1-999e97b6a09e.png)

**Comment by UmbralReaper** [comment]
> This has no relevance to BN currently, you're talking about a feature that was only broken in DDA

Yes. I was referring to the fact that the OP mentioned it as a feature that existed in DDA but not in BN. As the bug has been fixed in DDA, it is no longer a feature that exists in DDA but not in BN.

**Comment by Firestorm01X2** [comment]
List updated

**Comment by matshou** [comment]
I think we should pin this issue for more traction:
https://docs.github.com/en/issues/tracking-your-work-with-issues/pinning-an-issue-to-your-repository

**Comment by chaosvolt** [comment]
Two more for the list:
* Crafting menu shows amount of target product in crafting radius: https://github.com/CleverRaven/Cataclysm-DDA/pull/43393
* Use best tool in crafting radius when butchering: https://github.com/CleverRaven/Cataclysm-DDA/pull/48930

**Comment by chaosvolt** [comment]
Another one to look into, as turns out this is needed for mods to add their own custom power storage CBMs: https://github.com/CleverRaven/Cataclysm-DDA/pull/47777

Related improvements/fixes:
https://github.com/CleverRaven/Cataclysm-DDA/pull/49731
https://github.com/CleverRaven/Cataclysm-DDA/pull/51266
https://github.com/CleverRaven/Cataclysm-DDA/pull/51331

**Comment by chaosvolt** [comment]
Another suggestion that came up: https://github.com/CleverRaven/Cataclysm-DDA/pull/49330

In a nutshell, it would make it so that having a high effective dodge doesn't translate into bonus dodges, meaning once you've made your available dodges you don't get any more dodges. Suggestion came from BigBrawnyBodybuilderWeightlifter on the discord, with the main reasoning being it will help increase challenge for extremely high-dexterity/skill characters.

Main question is: in BN, you can still block repeatedly after you've run out of dodges, right? I vaguely recall DDA might make it so having no stamina negates blocking entirely but if I recall that isn't a thing in BN, right? If true, then such a change will make blocking come back into play even for high-dodge characters.

Another factor now that I think about it: this also means that bonus dodges from martial arts are currently almost irrelevant with high dodge skill, as exceeding your dodge limit just means you continue to dodge at a slight penalty until the hits finally start to accumulate, assuming the code in BN matches what the PR shows it was changed from.

Other alternatives and/or complimentary ideas we could implement:
1. Making it so having Uncanny Dodge active ignore this feature.
2. Allowing dodge skill to grant bonus dodges at certain skill thresholds to offset this.
3. There's been some talk of changes to how dodging functions that may or may not obsolete this idea.
4. Instead making the dodge skill penalties past your allowed dodges ramp up faster than normal instead of no dodging at all.
5. Instead making it so exceeding your dodge limit immediately becomes a fixed fraction of your effective dodge or dodge skill, whichever is (higher?)/(lower?).
6. Capping effective dodge or changing how it scales so you either can't become an untouchable dodging machine, or so you at least need outright superhuman dexterity and/or dodge-increasing mutations to become untouchable.
7. Implementing some form of dodge penalty that takes effect when facing more than one enemy in melee at the same time, that isn't dependent on a grab making it through your infinite dodging first.

**Comment by chaosvolt** [comment]
So, I was recently made aware of a PR we could potentially port over: https://github.com/CleverRaven/Cataclysm-DDA/pull/60039

This seems to allow for not only customizing exactly how many charges an item consumes on transform, but also allows for consumable tools. Thoughts?

**Comment by chaosvolt** [comment]
@Lamandus has asked (read: spammed my DMs >:V) that I port over https://github.com/CleverRaven/Cataclysm-DDA/pull/47209 to show volumes in the refill menu, but a lot of the file changes touch stuff that seems to be specific to nested containers. I don't see any easy way to grab this PR and convert it to work with BN's code.

First issue I've run afoul of is some easy to ignore divergences due to https://github.com/CleverRaven/Cataclysm-DDA/pull/42641, but then I run into not being sure what the BN equivalent to `max_containable_volume` is.

**Comment by chaosvolt** [comment]
Another thing that came up on the discord: https://github.com/CleverRaven/Cataclysm-DDA/pull/54022 which adds a location with some minor NPCs and associated missions.

Looks to be pure JSON stuff, so aside from checking for divergences in JSON it'd be easy but time-consuming to port over.

**Comment by shaneyao32** [comment]
The Exp version added a new sidebar interface calles structured layout, which is quite useful. Is there a chance to port it into BN？
https://github.com/CleverRaven/Cataclysm-DDA/pull/59182

**Comment by Krwak** [comment]
Better trading UI:
https://github.com/CleverRaven/Cataclysm-DDA/pull/52572

Better NPC dialogue UI:
https://github.com/CleverRaven/Cataclysm-DDA/pull/52486

**Comment by WreckingAbandon** [comment]
Some jsonified cbm effects, like Targeting System and the Internal Climate Control one.

**Comment by DrPariah** [comment]
Eternal Day/Night: [Eternal Night & Eternal Day options #53321](https://github.com/CleverRaven/Cataclysm-DDA/pull/53321)
Eternal Weather: [Eternal weather #59707](https://github.com/CleverRaven/Cataclysm-DDA/pull/59707)

Both of these look like fun options.

**Comment by silverwing235** [comment]
Fun...and broken links, as well.

**Comment by DrPariah** [comment]
> Fun...and broken links, as well.

Edited links they should be working now. Gotta admit, a Fimbulwinter setup (eternal night, eternal winter, eternal snowstorm) has a lot of appeal.

**Comment by TheEaterOfBeans** [comment]
Wind turbines aren't placeable like solar panels are.

**Comment by Zlorthishen** [comment]
> Eternal Day/Night: [Eternal Night & Eternal Day options #53321](https://github.com/CleverRaven/Cataclysm-DDA/pull/53321) Eternal Weather: [Eternal weather #59707](https://github.com/CleverRaven/Cataclysm-DDA/pull/59707)
> 
> Both of these look like fun options.

there should be a way to bring back acid rain

**Comment by DrPariah** [comment]
> > Eternal Day/Night: [Eternal Night & Eternal Day options #53321](https://github.com/CleverRaven/Cataclysm-DDA/pull/53321) Eternal Weather: [Eternal weather #59707](https://github.com/CleverRaven/Cataclysm-DDA/pull/59707)
> > Both of these look like fun options.
> 
> there should be a way to bring back acid rain

Japanese Variant had both Acid Rain AND Acid Snow, neither being too badly broken. Probably not too hard to port them for someone with actual coding skills. EDIT: here's the link to the variant.

[https://github.com/roloa/Cataclysm-DDA_variant](https://github.com/roloa/Cataclysm-DDA_variant)

**Comment by scarf005** [comment]
looks like a good performance port: https://github.com/CleverRaven/Cataclysm-DDA/pull/63056

**Comment by DrPariah** [comment]
[https://github.com/CleverRaven/Cataclysm-DDA/pull/56045](https://github.com/CleverRaven/Cataclysm-DDA/pull/56045)
[https://github.com/CleverRaven/Cataclysm-DDA/pull/56203](https://github.com/CleverRaven/Cataclysm-DDA/pull/56203)

Neat one from last year, it's a change to the music code that lets you designate playlists for things like the radio or such. Second one was a fix/update for the first.

**Comment by DrPariah** [comment]
[https://github.com/CleverRaven/Cataclysm-DDA/pull/49800](https://github.com/CleverRaven/Cataclysm-DDA/pull/49800)

NPCs able to use attack spells, I could see multiple uses for this.

**Comment by Zireael07** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/pull/65478

Recent change worth porting

**Comment by Lamandus** [comment]
> [CleverRaven/Cataclysm-DDA#65478](https://github.com/CleverRaven/Cataclysm-DDA/pull/65478)
> 
> Recent change worth porting

A lot of pseudo stuff going on, and the new forest floor thing doesn't seem to add a lot other than bloat

**Comment by SurFlurer** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/pull/57193 Enjoy the advancement of battery technology.

**Comment by DrPariah** [comment]
[https://github.com/WhiteCloud0123/CDDA-Breeze/pull/115](https://github.com/WhiteCloud0123/CDDA-Breeze/pull/115)

Not from mainline C:DDA, this is from the CDDA Breeze branch. Allows mounts to go up and down stairs. I don't think we have this.

**Comment by DrPariah** [comment]
[https://github.com/CleverRaven/Cataclysm-DDA/pull/62439](https://github.com/CleverRaven/Cataclysm-DDA/pull/62439)
[https://github.com/CleverRaven/Cataclysm-DDA/pull/62408](https://github.com/CleverRaven/Cataclysm-DDA/pull/62408)

There was a casual question in the Discord that got me to look these up. The top one is about additional traits for NPCs that make sure they don't bash things en route to pathing to them, returning to their start POS once they do, and an additional one for allowing NPCs to just plain not bother to investigate weird sounds at all.

The second one is for NPCs just plain not moving at all. This isn't near as useful for the intended goal (keeping NPCs from breaking everything and anything in their path en route to find out who said something just outside of their field of view), because it causes some other undesired behaviors (NPCs who don't fight back at all when attacked is one of them), but it seems like it could have some other very useful applications.

**Comment by DrPariah** [comment]
[https://github.com/WhiteCloud0123/CDDA-Breeze/pull/151](https://github.com/WhiteCloud0123/CDDA-Breeze/pull/151)

Monsters that can equip weapons, from the Breeze fork of C:DDA.

**Comment by chaosvolt** [comment]
Two things Boophis brought up in DMs:
1. https://github.com/CleverRaven/Cataclysm-DDA/pull/49800 (spellcasting NPC extension)
2. https://github.com/CleverRaven/Cataclysm-DDA/pull/51981 and https://github.com/CleverRaven/Cataclysm-DDA/pull/52381 (playing games with NPC companions, relevant items may need to be ported over too)

**Comment by DrPariah** [comment]
[https://github.com/CleverRaven/Cataclysm-DDA/pull/51739](https://github.com/CleverRaven/Cataclysm-DDA/pull/51739)

This looks like a VERY useful PR, allowing conversations with monsters and (most importantly) items. This would allow us to use maps/journals/etc. for quests without having to use a SPAWN_NPC item (though that is still a viable option and very appropriate in certain cases like the holograms).

**Comment by chaosvolt** [comment]
Possible performance improvement, thanks to @Maleclypse for bringing my attention to it in the discord: https://github.com/CleverRaven/Cataclysm-DDA/pull/67403

**Comment by scarf005** [comment]
would be great for modding.
https://github.com/CleverRaven/Cataclysm-DDA/pull/66623

btw we should really make use of tasklist and project board someday to track what's done and undone.

**Comment by ghost** [comment]
Nested crafting categories could condense the crafting menu a ton especially for mods that add a lot of similar items like extra medieval weaponry. https://github.com/CleverRaven/Cataclysm-DDA/pull/59924
